### PR TITLE
implement text based entity card/content filters

### PR DESCRIPTION
**Pull request by freben**
## Overview

This PR introduces the ability for card/content extensions to the catalog frontend plugin, to receive a `filter` argument that's text based, in addition to the previous callback based system. The filter decides whether the extension should be rendered or not. This lets you for example only show the CI/CD tab for a select set of kinds - even if you are using custom kinds. All from the convenience of your app-config file.

Example in the configuration file for declarative integration:

```yaml
app:
  extensions:
    - entity.cards.about
    - entity.cards.links:
        config:
          filter: 'kind:component has:links'
```

The suggested format in this PR is a space separated list of `key:parameters` pairs, where all such pairs need to match an entity for it to be rendered. Additionally, the parameters list is comma separated, letting you for example target multiple kinds with `kind:user,group`. This is heavily inspired by the github filter syntax for issues and pull requests.

## Future work

For now both the syntax and the features are fixed. You only have `kind:`, `type:`, `is:`, and `has:`, and the latter two have a set number of parameters that they support. In the future we may want to add the ability to inject custom additions to those.

## Alternatives

This is, of course, a bespoke format. We went back and forth a bit on how it should look. It needs to be convenient to type for humans, and still be easy enough for e.g. a management console to have a UI for and generate the string from.

One contrast that comes to mind, is the catalog backend API. It doesn't have a query syntax like this yet, and [its current filtering capabilities](https://backstage.io/docs/features/software-catalog/software-catalog-api#filtering) do leave some room for improvement. If we were to introduce a query format for the backend, it would be good if it could tap into the same syntax as we choose here. Or put differently, the syntax chosen here is knowingly different from the catalog backend filters; is that a bad thing or does this actually make some movement toward a better way of expressing filters there too in the future?

Open to suggestions and ideas.

### CODE REVIEW

**Comment by freben** [comment]
One thing of note here is that the filter builder throws errors. This does probably fail "too hard", leading to error overlays if you have bad syntax problems. We may want to compile forgivingly (making filters more permissive in the face of failures, ie ignoring restriction clauses) and emitting warnings instead.

File: plugins/catalog/src/alpha/plugin.tsx [hunk 1]
```diff
@@ -45,6 +46,7 @@ export default createPlugin({
     viewTechDoc: convertLegacyRouteRef(viewTechDocRouteRef),
     createComponent: convertLegacyRouteRef(createComponentRouteRef),
     createFromTemplate: convertLegacyRouteRef(createFromTemplateRouteRef),
+    unregister: convertLegacyRouteRef(unregisterRedirectRouteRef),
```
Comment (Rugvip) [inline plugins/catalog/src/alpha/plugin.tsx:49]: https://github.com/backstage/backstage/pull/21481/files#diff-8172f1eed894166ce8329a218913d4b4b6dc47a3357ec8e6d404b9774c4dabf6 😅, I'll extract it, https://github.com/backstage/backstage/pull/21495

**Review by Rugvip — CHANGES_REQUESTED** [review]
Nice 😎

Main tricky bit for me is the interaction between plugin builder and integrator when it comes to filter definitions, there's a comment starting up that discussion. Rest is mostly nits 😅

File: plugins/catalog-common/alpha-api-report.md [hunk 2]
```diff
@@ -38,6 +39,26 @@ export const catalogPermissions: (
   | ResourcePermission<'catalog-entity'>
 )[];
 
+// @alpha
+export function parseFilterExpression(
```
[the remark this answers is not in the document]
Comment (freben) [reply plugins/catalog-common/alpha-api-report.md:43]: Made it internal!

File: plugins/catalog-common/alpha-api-report.md [hunk 3]
```diff
@@ -38,6 +39,26 @@ export const catalogPermissions: (
...
+        expression: string;
+      },
+    ) => void;
+    onEvaluateError?: (
+      error: Error,
+      context: {
+        expression: string;
+        entity: Entity;
+      },
+    ) => void;
```
[the remark this answers is not in the document]
Comment (freben) [reply plugins/catalog-common/alpha-api-report.md:58]: yah, returned an array instead

File: plugins/catalog-react/src/alpha.tsx [hunk 4]
```diff
@@ -25,54 +24,23 @@ import {
...
 
 /** @alpha */
 export const entityContentTitleExtensionDataRef =
   createExtensionDataRef<string>('plugin.catalog.entity.content.title');
 
 /** @alpha */
 export const entityFilterExtensionDataRef = createExtensionDataRef<
-  (ctx: { entity: Entity }) => boolean
+  string | ((entity: Entity) => boolean)
 >('plugin.catalog.entity.filter');
```
[the remark this answers is not in the document]
Comment (freben) [reply plugins/catalog-react/src/alpha.tsx]: alright, done!

File: plugins/catalog-common/src/filter/matrchers/createIsMatcher.ts [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+ * Matches on different semantic properties of the entity
+ */
+export function createIsMatcher(
+  parameters: string[],
+  onParseError: (error: Error) => void,
+): EntityMatcherFn {
+  const allowedMatchers: Record<string, EntityMatcherFn> = {
+    orphan: entity =>
+      Boolean(entity.metadata.annotations?.['backstage.io/orphan']),
+  };
```
[the remark this answers is not in the document]
Comment (freben) [reply plugins/catalog-common/src/filter/matrchers/createIsMatcher.ts]: sure! done

File: plugins/catalog-common/src/filter/matrchers/createHasMatcher.ts [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+  const allowedMatchers: Record<string, EntityMatcherFn> = {
+    labels: entity => {
+      return Object.keys(entity.metadata.labels ?? {}).length > 0;
+    },
+    annotations: entity => {
+      return Object.keys(entity.metadata.annotations ?? {}).length > 0;
+    },
+    links: entity => {
+      return (entity.metadata.links ?? []).length > 0;
+    },
```
[the remark this answers is not in the document]
Comment (freben) [reply plugins/catalog-common/src/filter/matrchers/createHasMatcher.ts]: Or `annotation:github.com/project-slug` or `field:metadata.annotations.github.com/project-slug` perhaps. `has` was meant to be an easy-to-type, quick and "semantic" set of common utilities that you want for your entities. So with that philosophy, you'd add a custom `has` extension, perhaps making it `has:github-project-slug`? Or `is:github-enabled` or `has:github-annotations`? But yeah, the question becomes how plugins can provide those, then.

**Review by Rugvip — APPROVED** [review]
:shipit:

**Pull request merged**